pkgs/development/compilers/dotnet/patch-nupkgs.nix

```diff
@@ -31,45 +31,54 @@ writeShellScriptBin "patch-nupkgs" (
   ''
     set -euo pipefail
     shopt -s nullglob
-    isELF() {
-        local fn="$1"
-        local fd
-        local magic
-        exec {fd}< "$fn"
-        read -r -n 4 -u "$fd" magic
-        exec {fd}<&-
-        if [ "$magic" = $'\177ELF' ]; then return 0; else return 1; fi
-    }
     cd "$1"
   ''
   + lib.optionalString stdenv.hostPlatform.isLinux ''
+    isNonStaticExecutableELF() {
+      local magic
+      local e_type
+      local interpreter
+
+      if magic=$(od --address-radix=n --read-bytes=4 --format=c "$1") \
+        && [ "$magic" = " 177   E   L   F" ] \
+        && e_type=$(od --address-radix=n --skip-bytes=0x10 --read-bytes=2 --format=x2 "$1") \
+        && interpreter=$(${patchelf}/bin/patchelf --print-interpreter "$1") \
+        && [ "$e_type" != " 0002" -o -n "$interpreter" ]; then return 0; else return 1; fi
+    }
+
     for x in */* */*; do
       # .nupkg.metadata is written last, so we know the packages is complete
       [[ -d "$x" ]] && [[ -f "$x"/.nupkg.metadata ]] \
         && [[ ! -f "$x"/.nix-patched ]] || continue
       echo "Patching package $x"
       find "$x" -type f -print0 | while IFS= read -rd "" p; do
         if [[ "$p" != *.nix-patched ]] \
-          && isELF "$p" \
-          && ${patchelf}/bin/patchelf --print-interpreter "$p" &>/dev/null; then
+          && isNonStaticExecutableELF "$p" \
+          && interpreter=$(${patchelf}/bin/patchelf --print-interpreter "$p") \
+          && [ -n "$interpreter" ]; then
           tmp="$p".$$.nix-patched
           # if this fails to copy then another process must have patched it
           cp --reflink=auto "$p" "$tmp" || continue
           echo "Patchelfing $p as $tmp"
-          ${patchelf}/bin/patchelf \
-            --set-interpreter "${stdenv.cc.bintools.dynamicLinker}" \
-            "$tmp" ||:
-          # This makes sure that if the binary requires some specific runtime dependencies, it can find it.
-          # This fixes dotnet-built binaries like crossgen2
-          ${patchelf}/bin/patchelf \
-            --add-needed libicui18n.so \
-            --add-needed libicuuc.so \
-            --add-needed libz.so \
-            --add-needed libssl.so \
-            "$tmp"
-          ${patchelf}/bin/patchelf \
-            --add-rpath "${binaryRPath}" \
-            "$tmp" ||:
+
+          (
+            ${patchelf}/bin/patchelf \
+              --set-interpreter "${stdenv.cc.bintools.dynamicLinker}" \
+              "$tmp"
+
+            # This makes sure that if the binary requires some specific runtime dependencies, it can find it.
+            # This fixes dotnet-built binaries like crossgen2
+            ${patchelf}/bin/patchelf \
+              --add-needed libicui18n.so \
+              --add-needed libicuuc.so \
+              --add-needed libz.so \
+              --add-needed libssl.so \
+              "$tmp"
+            ${patchelf}/bin/patchelf \
+              --add-rpath "${binaryRPath}" \
+              "$tmp"
+          ) ||:
```

Comment:
This error handling is going to change the behaviour a bit.  I'm not sure if it's a problem, but previously, if `--set-interpreter` failed, `--add-needed` would still happen.

Unfortunately it's not clear why errors were being ignored in the first place.

Reply:
True, I think I might just disable the error ignoring and let it fail, instead of silently swallowing them.

Reply:
What happens without the `||:`.  Maybe it's safe now that we have the interpreter check?

If we do need it, I think it would be good to have a comment explaining why.

Reply:
I'm not sure what happens without it. I think it should be nothing, however I'm not sure.